core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -157,6 +163,115 @@ public void segmentsLost(IntSet lostSegments) {
       return (Function) ignoreSegmentsFunction;
    }
 
+   private <I, R> void handleParallelSegment(PrimitiveIterator.OfInt segmentIter, int initialSegment, CacheSet<I> set,
+         Set<K> keysToExclude, Function<I, K> toKeyFunction, Function<? super Publisher<I>, ? extends CompletionStage<R>> transformer,
+         FlowableProcessor<R> processor, IntSet concurrentSegments, SegmentListener listener) {
+      // Controls how many pending segment completions we have. We can't complete the result until
+      // all are complete. The invoking thread always has 1 for itself to ensure it is not completed early while
+      // submitting the tasks.
+      AtomicInteger pendingCompletions = new AtomicInteger(1);
+      // This variable determines if our processor was serialized or not - We use a non serialized processor
+      // for Publishers that are completed in the invoking thread. If a result was not completed immediately we have
+      // to convert our Processor to be serialized to ensure we are only calling onNext from 1 thread at a time
+      boolean serializedProcessor = false;
+      try {
+         while (true) {
+            // The first run initialSegment will be 0 or greater. We use that segment and and set it to -1 to notify
+            // our next run to try to steal a segment from the iterator to process until the iterator runs out
+            // of segments
+            int nextSegment;
+            if (initialSegment != -1) {
+               nextSegment = initialSegment;
+               initialSegment = -1;
+            } else {
+               nextSegment = getNextSegment(segmentIter);
+               if (nextSegment == -1) {
+                  break;
+               }
+            }
+
+            pendingCompletions.getAndIncrement();
+
+            Flowable<I> innerFlowable = Flowable.fromPublisher(set.localPublisher(nextSegment));
+            if (keysToExclude != null) {
+               innerFlowable = innerFlowable.filter(i -> !keysToExclude.contains(toKeyFunction.apply(i)));
+            }
+
+            CompletionStage<R> stage = transformer.apply(innerFlowable);
+            CompletableFuture<R> future = stage.toCompletableFuture();
+            // This will always be true if there isn't a store, however in most cases with a store this will
+            // be false as we have to wait until the store can publish all the entries (which is done asynchronously)
+            if (future.isDone() && !future.isCompletedExceptionally()) {
```

Comment:
I had a version where I didn't; I think the reason is because we invoke join below and I didn't want to check `isDone` and `isCompletedExceptionally` twice when this approach only does it once.

Reply:
Actually misthought, I can change this over.